### ISSUE TITLE
Discriminated union types better error reporting

### ISSUE THREAD

**Issue by whitecolor** [Microsoft/TypeScript#10867]
This the example from docs:
![image](https://cloud.githubusercontent.com/assets/736697/18444742/eef61048-7933-11e6-9eff-ba5d276f083f.png)

``` ts
interface Square {
    kind: "square";
    size: number;
}

interface Rectangle {
    kind: "rectangle";
    width: number;
    height: number;
}

interface Circle {
    kind: "circle";
    radius: number;
}

type Shape = Square | Rectangle | Circle;

let shape: Shape = {
  kind: "square",
  width: 10,
  height: 10
}
```

So we create object with type `Shape` of kind `square` and try to pass incorrect other props (for this kind), and TS emits the error that object is no assignable to last type that is in union (`Circle` in this case). I believe it would be better if TS would recoginze that we are trying to make `Square` type in this case and emitted corresponding error message.

**Comment by Arnavion** [comment]
There's another recent issue #10849 that talks about this and other discriminated unions issues. Perhaps one could be merged into the other?

**Comment by KavanShaban** [comment]
Yes, this is a subset of the issue #10849 that reported a few days back.

**Comment by RyanCavanaugh** [comment]
Work item for 2.2: Figure out how to detect discriminated unions and report errors based on a matched discriminant field

**Comment by sandersn** [comment]
Fix is up at #14006
